CMSgov/easi-app
# [EASI-2604, EASI-3239] IT gov admin action success/error and modal

# EASI-2604 & EASI-3239

## Changes and Description

### EASI-2604
- Added server error message to <ActionForm>
- Added success message to actions page on form submit
  - `successMessage` prop sets message text

### EASI-3239
- Added confirmation modal to <ActionForm>
  - Form submit / modal trigger logic: if `modal` prop is defined, modal is triggered on submit button click AFTER form field validation passes
- `modal` prop sets modal title and content (leave undefined if no modal)
  - `content` prop can be `string` or `React.ReactNode`. This makes it possible to pass a `<Trans>` component if needed (see example from [Request Edits modal](https://www.figma.com/file/ChzAP34A2DVvQUNQwD7lCt/IT-Governance-Next?type=design&node-id=2202%3A128126&mode=design&t=vUGDVIsj6h3w0fBU-1) below)

```jsx
<ActionForm
   modal={{
      title: t('requestEdits.modal.title'),
      content: (
        <Trans
          i18nKey="action:requestEdits.modal.content"
          components={{ p: <p /> }}
          values={{ formName: 'Draft Business Case' }}
        />
      )
   }}
/>
``` 

- Updated Modal component (src/components/Modal) close button style to match Figma, USWDS, and Trussworks component

### Architectural changes
- Form submit logic was moved from individual forms (request edits, next step, etc) to the `<ActionForm>` component
  - The `onSubmit` prop should **only** handle the mutation and actually completing the action
    - Field validation (RHF's `handleSubmit` function) and error/success handling is now done in the `<ActionForm>` component
  - new `onSubmit` prop function type:
    ```ts
    onSubmit: (formData: TFieldValues) => Promise<void>;
    ```



## How to test this change

- Sign in as IT Gov admin with `itGovV2Enabled` flag turned on
- Go to /governance-review-team/:systemIntakeId/actions
- Select "Request Edits"
- Submit form
  - Modal should trigger on form submit
  - Click "complete action" button on modal
  - Form should redirect back to main actions page and show success message
- To test error message, open `RequestEdits` component file (src/views/GovernanceReviewTeam/Actions/RequestEdits.tsx)
  - Uncomment line 24 to trigger error on form submit
  - Check that server error message appears on form

## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
